pkg/auth/auth.go

```diff
@@ -0,0 +1,290 @@
+	"github.com/ethersphere/bee/pkg/logging"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type authRecord struct {
+	Role   string    `json:"r"`
+	Expiry time.Time `json:"e"`
+}
+
+type Authenticator struct {
+	passwordHash []byte
+	ciph         *encrypter
+	enforcer     *casbin.Enforcer
+	log          logging.Logger
+}
+
+func New(encryptionKey, passwordHash string, logger logging.Logger) (*Authenticator, error) {
+	m, err := model.NewModelFromString(`
+	[request_definition]
+	r = sub, obj, act
+
+	[policy_definition]
+	p = sub, obj, act
+
+	[policy_effect]
+	e = some(where (p.eft == allow))
+
+	[matchers]
+	m = r.sub == p.sub && (keyMatch(r.obj, p.obj) || keyMatch(r.obj, '/v1'+p.obj)) && regexMatch(r.act, p.act)`)
+
+	if err != nil {
+		return nil, err
+	}
+
+	e, err := casbin.NewEnforcer(m)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := applyPolicies(e); err != nil {
+		return nil, err
+	}
+
+	ciph, err := newEncrypter([]byte(encryptionKey))
+	if err != nil {
+		return nil, err
+	}
+
+	auth := Authenticator{
+		enforcer:     e,
+		ciph:         ciph,
+		passwordHash: []byte(passwordHash),
+		log:          logger,
+	}
+
+	return &auth, nil
+}
+
+func (a *Authenticator) Authorize(password string) bool {
+	return nil == bcrypt.CompareHashAndPassword(a.passwordHash, []byte(password))
+}
+
+var ErrExpiry = errors.New("expiry duration must be a positive number")
+
+func (a *Authenticator) GenerateKey(role string, expiryDuration int) (string, error) {
+	if expiryDuration == 0 {
+		return "", ErrExpiry
+	}
+
+	ar := authRecord{
+		Role:   role,
+		Expiry: time.Now().Add(time.Second * time.Duration(expiryDuration)),
+	}
+
+	data, err := json.Marshal(ar)
+	if err != nil {
+		return "", err
+	}
+
+	encryptedBytes, err := a.ciph.encrypt(data)
+	if err != nil {
+		return "", err
+	}
+
+	apiKey := base64.StdEncoding.EncodeToString(encryptedBytes)
+
+	return apiKey, nil
+}
+
+var ErrTokenExpired = errors.New("token expired")
+
+func (a *Authenticator) RefreshKey(apiKey string, expiryDuration int) (string, error) {
+	if expiryDuration == 0 {
+		return "", ErrExpiry
+	}
+
+	decoded, err := base64.StdEncoding.DecodeString(apiKey)
+	if err != nil {
+		return "", err
+	}
+
+	decryptedBytes, err := a.ciph.decrypt(decoded)
+	if err != nil {
+		return "", err
+	}
+
+	var ar authRecord
+	if err := json.Unmarshal(decryptedBytes, &ar); err != nil {
+		return "", err
+	}
+
+	if time.Now().After(ar.Expiry) {
+		return "", ErrTokenExpired
+	}
+
+	ar.Expiry = time.Now().Add(time.Duration(expiryDuration) * time.Second)
+
+	data, err := json.Marshal(ar)
+	if err != nil {
+		return "", err
+	}
+
+	encryptedBytes, err := a.ciph.encrypt(data)
+	if err != nil {
+		return "", err
+	}
+
+	apiKey = base64.StdEncoding.EncodeToString(encryptedBytes)
+
+	return apiKey, nil
+}
+
+func (a *Authenticator) Enforce(apiKey, obj, act string) (bool, error) {
+	decoded, err := base64.StdEncoding.DecodeString(apiKey)
+	if err != nil {
+		a.log.Error("decode token", err)
```

Comment:
I would avoid both logging and returning errors in this function.

Reply:
I'm going to reduce logging in a future PR but right now we need detailed information when certain cases occur, like token expiration.